src/app/feature/nav-stack/nav-stack.service.ts

```diff
@@ -0,0 +1,83 @@
+import { Injectable } from "@angular/core";
+import { ModalController } from "@ionic/angular";
+import { INavStackConfig, NavStackComponent } from "./components/nav-stack/nav-stack.component";
+import { SyncServiceBase } from "src/app/shared/services/syncService.base";
+
+interface NavStackModal extends HTMLIonModalElement {}
+
+@Injectable({
+  providedIn: "root",
+})
+export class NavStackService extends SyncServiceBase {
+  private openNavStacks: HTMLIonModalElement[] = [];
+
+  constructor(private modalCtrl: ModalController) {
+    super("navStack");
+    // NB: Actions are registered in nav-stack module
+  }
+
+  /**
+   * Create and present a new nav-stack modal and push it to openNavStacks array.
+   * Await and remove from openNavStacks array on dismiss
+   */
+  public async pushNavStack(navStackConfig: INavStackConfig) {
+    const modal = await this.createNavStackModal(navStackConfig);
+    await this.presentAndTrackModal(modal);
+  }
+
+  public async closeAllNavStacks() {
+    // Close nav-stacks in reverse order
+    for (let index = this.openNavStacks.length - 1; index >= 0; index--) {
+      await this.closeNavStack(index);
+    }
```

Comment:
nit(non-blocking): I'm not sure if this level of UX is necessary/helpful... if closing all stacks I don't think we necessarily need to do them one at a time and in order, might be better to just use the modalCtrl to close all modals and reset the array manually (if not automated) - or map the items to a Promise.all so they can work in parallel

Reply:
I found that if I do, for example, 
```ts
  public async closeAllNavStacks() {
    await Promise.all(this.openNavStacks.map((_, index) => this.closeNavStack(index)));
  }
```
Then the test for `closeAllNavStacks()` fails, with the length of the `openNavStacksArray` still being 1 (even with an artificial `_wait` time added to the test. I'm not quite sure what this is due to, potentially just something in the test setup, but I haven't been able to resolve it with various async configurations, so this brute force approach seemed acceptable.

Reply:
Ah, the problem is for those tests we were still manually adding the navStack modal to the array and calling dismiss, instead of programmatically generating them via the `pushNavStack` method (which also adds the event bindings for dismiss). I think part of the reason for this is to have direct access to the modal spy that would have been generated.

I've added [a50d10c](https://github.com/IDEMSInternational/open-app-builder/pull/2550/commits/a50d10c151de037d430cb3cd69580d28e8a62129) to fix this - now the `pushNavStack` method returns the created modal (to make it easier to keep reference to it), and changed the service code to first push the nav stacks before dismissing. 

I've also updated the promise to call the specific modal `dismiss` function instead of looking up its index (the `openNavStacks` already store the full modal so no need to retrieve). Look to be working for me now, so if you're happy I'd say good to go pending any input from @esmeetewinkel